x/slashing/keeper.go

```diff
@@ -120,3 +130,36 @@ func (k Keeper) handleValidatorSignature(ctx sdk.Context, pubkey crypto.PubKey,
 	// Set the updated signing info
 	k.setValidatorSigningInfo(ctx, address, signInfo)
 }
+
+// AddValidators adds the validators to the keepers validator addr to pubkey mapping.
+func (k Keeper) AddValidators(vals []abci.Validator) {
+	for i := 0; i < len(vals); i++ {
+		val := vals[i]
+		pubkey, err := tmtypes.PB2TM.PubKey(val.PubKey)
+		if err != nil {
+			continue
+		}
+		k.addPubkey(pubkey, false)
+	}
+}
+
+// TODO: Make a method to remove the pubkey from the map when a validator is unbonded.
+func (k Keeper) addPubkey(pubkey crypto.PubKey, del bool) {
+	addr := new([tmhash.Size]byte)
+	copy(addr[:], pubkey.Address())
+	if del {
+		delete(k.addressToPubkey, *addr)
+	} else {
+		k.addressToPubkey[*addr] = pubkey
+	}
+}
+
+func (k Keeper) getPubkey(address crypto.Address) (crypto.PubKey, error) {
+	addr := new([tmhash.Size]byte)
```

Comment:
Is there a reason we allocate a pointer instead of just:
```go
var addr [tmhash.Size]byte
copy(addr[:], address)
// ...
```

Reply:
`new` returns a memory address -- that way you don't have to do `*addr` but just `addr`. No big deal though.